src/lib.rs

```diff
@@ -253,30 +260,62 @@ where
 /// the cookie is overwritten by a new message.
 /// ```no_run
 /// # use actix_web_flash::{FlashMiddleware};
-/// # use actix_web::{App, server};
-/// server::new(|| {
+/// # use actix_web::{App, HttpServer};
+/// HttpServer::new(|| {
 ///     App::new()
-///         .middleware(FlashMiddleware::default())
+///         .wrap(FlashMiddleware::default())
 /// }).bind("127.0.0.1:8080")
 ///     .unwrap()
 ///     .run();
 /// ```
-pub struct FlashMiddleware();
+pub struct FlashMiddleware;
 
-impl<S> Middleware<S> for FlashMiddleware {
-    fn response(
-        &self,
-        req: &HttpRequest<S>,
-        mut resp: HttpResponse,
-    ) -> Result<Response, actix_web::Error> {
-        let mut jar = CookieJar::new();
-        if let Some(cookie) = req.cookie(FLASH_COOKIE_NAME) {
-            jar.add_original(cookie);
-            jar.remove(Cookie::named(FLASH_COOKIE_NAME));
-        }
-        for cookie in jar.delta() {
-            resp.add_cookie(cookie)?;
-        }
-        Ok(Response::Done(resp))
+impl<S, B> Transform<S> for FlashMiddleware
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type InitError = ();
+    type Transform = FlashMiddlewareServiceWrapper<S>;
+    type Future = FutureResult<Self::Transform, Self::InitError>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        fut_ok(FlashMiddlewareServiceWrapper(service))
+    }
+}
+
+/// The actual Flash middleware
+pub struct FlashMiddlewareServiceWrapper<S>(S);
+
+impl<S, B> Service for FlashMiddlewareServiceWrapper<S>
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = Box<Future<Item = Self::Response, Error = Self::Error>>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.0.poll_ready()
+    }
+
+    fn call(&mut self, req: ServiceRequest) -> Self::Future {
+        Box::new(self.0.call(req).map(move |mut res| {
+            let mut jar = CookieJar::new();
+            if let Some(cookie) = res.request().cookie(FLASH_COOKIE_NAME) {
+                jar.add_original(cookie);
+                jar.remove(Cookie::named(FLASH_COOKIE_NAME));
+            }
+            for cookie in jar.delta() {
+                let _ = res.response_mut().add_cookie(cookie);
```

Comment:
Swallowing this error seems like a bad idea. This might be a bit annoying to do with the future but I think this should definitely be handled.